### ISSUE TITLE
Can we have a discussion about the Go code generator?

### ISSUE THREAD

**Issue by arekkas** [swagger-api/swagger-codegen-generators#34]
So I saw that 3.0.0 will use this repository for codegen generators. Version 2.3.x introduced some major breaking changes to the produced code. While most of these changes made sense, some of them are - in my opinion - bad choices.

I would like to use the chance that releasing 3.0.0 is, to start a discussion on providing a better code generator for Golang.

I'd like to start with some things that I think are really well done in the new 2.3.x release:

1. It's now possible to provide a custom `http.Client`
2. It's now possible provide a `context.Context` in API calls

I think there are also some issues.

1. Redirects do apparently no longer work, because all status codes higher than 299 result in an error:
```
	if localVarHttpResponse.StatusCode >= 300 {
		bodyBytes, _ := ioutil.ReadAll(localVarHttpResponse.Body)
		return localVarHttpResponse, reportError("Status: %v, Body: %s", localVarHttpResponse.Status, bodyBytes)
	}
```
2. Adding request bodies now requires passing a `map[string]interface{}` that needs, for example, a `body` key.

The second limitation is especially painful from an API point of view as type assertions are necessary to decode the payload, and the API is not very developer friendly in the context of Go. In real-world use it looks something like this:

```
CreateResource(context.Context, map[string]interface{}{"body": &SomeStruct{}})
```

I propse that instead, a type is used to solve this, for example:

```
type CreateResourceParameters struct {
  Body *SomeStruct
  Path ...
  Query ...
}
```

This would make the process of passing down request parameters much more usable and less error prone.

Let me know what you think.

**Comment by webron** [comment]
@antihax, @bvwells - your input would be appreciated.

**Comment by arekkas** [comment]
One point I forgot is error payloads. In the spec, it's possbile to provide an error payload for e.g. 401 or 403. With the current `swagger-codegen` template for Go, the response body is actually not readable from outside as the body is closed within the generated code.

What I suggest is to have custom errors which implement the `error.Error` interface instead. These error structs are hydrated depending on the status code. The developer may use type assertion to get the full error message, including the original body.

**Comment by antihax** [comment]
> Redirects do apparently no longer work, because all status codes higher than 299 result in an error:

This should be fixable.

> Adding request bodies now requires passing a map[string]interface{} that needs, for example, a body key.

I agree this was a bad design choice, but there are limited options available to handle optional parameters. 

> What I suggest is to have custom errors which implement the error.Error interface instead. These error structs are hydrated depending on the status code.

I know the error models are generated, so just a matter of populating it. This may confuse some folks getting an `interface{}` back instead of `error`. Perhaps we extend the response instead with the struct?

**Comment by arekkas** [comment]
> I know the error models are generated, so just a matter of populating it. This may confuse some folks getting an interface{} back instead of error. Perhaps we extend the response instead with the struct?

You can type assert errors, no need for an interface.

> I agree this was a bad design choice, but there are limited options available to handle optional parameters.

Optional parameters are solvable with structs, as no field in the struct is required!

If something isn't clear I'm happy to explain a bit more.

**Comment by antihax** [comment]
> Optional parameters are solvable with structs, as no field in the struct is required!

An integer will default to 0, a boolean will default to false both could be legitimate values. How do you determine if they are optional or not?

**Comment by arekkas** [comment]
> An integer will default to 0, a boolean will default to false both could be legitimate values. How do you determine if they are optional or not?

I see - where is this currently an issue though? As far as I understood it (I played around with the new code for ~ hour), we're still using structs for e.g. body payloads. There we can't define empty parts either, right? One possibility is, of course, to use pointers. Using pointers for simple values however is not that nice.

**Comment by antihax** [comment]
Indeed, pointers would require you to copy to a new variable before calling as a goroutine or you will race on the variable.

Unfortunately the API i am mostly working with is littered with such things.

In an update request this is optional. If you supply true or false, it will update the entry. So if it was true and Go defaults it to false, it will update it when you do not wish it to.

```
         {
            "name": "watched",
            "in": "query",
            "description": "Whether the new contact should be watched, note this is only effective on characters",
            "required": false,
            "type": "boolean",
          },
```

**Comment by arekkas** [comment]
Ok I see. I think there are multiple things we could consider here:

1. Body payloads currently use JSON marshalling from structs. This implies that your use case would actually not work if that was in the body, right?
2. Query parameters *could* be solves using `url.Values{}` which is the stdlib variant for defining query parameters. I understand that this will not help us in replacing the `map[string]interface{}` with a struct, but it will at least use the stdlib which people are used to in this context!

The third part is, I think, the path. I'm not sure if all parameters in the path are always required or not. Currently, the path uses (if I understand it correctly) the methods arguments. So here, in the current version of the API, this wouldn't work either, right?

In conclusion I think that we could probably leverage `url.Values{}` to use the "standard way" for defining queries, and body structs for the request body. However, it won't solve this issue for your use case, if the payload is in the body and not in the query.

**Comment by antihax** [comment]
It is definitely worth exploring. Could you not convert `url.Values{}` to JSON/XML for the body?

**Comment by arekkas** [comment]
> It is definitely worth exploring. Could you not convert url.Values{} to JSON/XML for the body?

No, `url.Values` only supports strings: `map[string][]string{}`

**Comment by antihax** [comment]
That looks like the way to go. We would get compile time type checking back.

```
type StringType struct{
    val string
    set bool
}
func String(v string) StringType {
    return StringType{v, true}
}
func (s StringType) IsSet() bool {
    return s.set
}
func (s StringType) Value() string {
    return s.String()
}
func (s StringType) String() string {
    if !s.set {
        return ""
    }
    return s.val
}
```

**Comment by arekkas** [comment]
So the question is, when a value is not required, if it should be nullable per default or not (as a Go developer I would say no, it's not nullable per default). Is there a way to express *nullability* in the swagger spec?

**Comment by arekkas** [comment]
The segment Go SDK uses composable types:

```
client.Enqueue(analytics.Identify{
  UserId: "019mr8mf4r",
  Traits: analytics.NewTraits().
    SetName("Michael Bolton").
    SetEmail("mbolton@initech.com").
    Set("plan", "Enterprise").
    Set("friends", 42),
})
```

I don't really like this approach because doesn't allow marshal/unmarshal.

**Comment by arekkas** [comment]
Marshall seems to work quite well with nullable values, check [this](https://play.golang.org/p/o0LF1f4ky5a):

```go
package main

import (
	"encoding/json"
	"fmt"
)

type Foo struct {
	Foo *string `json:"foo"`
}

func main() {
	out, err := json.Marshal(&Foo{})
	if err != nil {
		panic(err)
	}
	fmt.Printf("Hello, playground %s", out)
}
```

```
Hello, playground {"foo":null}
Program exited.
```

**Comment by antihax** [comment]
I agree with the AWS way also, but let me know what you find.

**Comment by arekkas** [comment]
Another idea would be to have something like this:

```
type MyRequestPayload struct {
  Body SomeType
  Query SomeType1
  Path SomeType2

  BodyCustom map[string]interface{}
  QueryCustom map[string]interface{}
  PathCustom map[string]interface{}
}
```

**Comment by webron** [comment]
OAS2 indeed doesn't support `null` values directly.
OAS3 - does, using the `nullable` property.

**Comment by arekkas** [comment]
This could actually be improved:

```
type MyRequestPayload struct {
  *MyRequestPayloadNotNullable 
  *MyRequestPayloadNullable 
}

type MyRequestPayloadNotNullable struct {
  Body SomeType
  Query SomeType1
  Path SomeType2
}

type MyRequestPayloadNullable struct {
  BodyCustom map[string]interface{}
  QueryCustom map[string]interface{}
  PathCustom map[string]interface{}
}

var payload MyRequestPayload = NewMyRequestPayload()
// or if it should be nullable
var payload MyRequestPayload = NewMyRequestPayloadNullable()

client.MyRequest(context, payload)
```

**Comment by arekkas** [comment]
Ok that's quite unfortunate, because primitive null values are simply not standard for Go (see the sql.StringNull confusion/fiasko) :(

I checked out the google appengine SDK, but it doesn't seem to have nullable values.

**Comment by arekkas** [comment]
It seems like the Google Cloud SDK supports nullable values, it has also custom types: https://godoc.org/cloud.google.com/go/bigquery#NullInt64

```
type NullInt64 struct {
	Int64 int64
	Valid bool // Valid is true if Int64 is not NULL.
}

func (n NullInt64) String() string { return nullstr(n.Valid, n.Int64) }
```

So if even Google does this in Go, it does seem like it's a good idea.

**Comment by bvwells** [comment]
@arekkas - issue #7809 has been logged along the same lines as the first issue you discuss. Any comments on the proposed PR?

**Comment by arekkas** [comment]
Thank you for pointing me to it, I commented and discourage merging it as it will introduce serious issues to anyone not paying attention.

**Comment by arekkas** [comment]
My proposal for improving the API is as followed.

```
type NilString *string
// func (s NilString) String() string ...


type SomeRequestBody struct {
  Param1 NilString `json:"param1,omitempty"`
  Param2 string `json:"param2"`
  // ...
}
type SomeRequestPath struct {
  Param3 NilString `json:"param3,omitempty"`
  Param4 string `json:"param4"`
  // ...
}
type SomeRequestQuery struct {
  Param5 NilString `json:"param5,omitempty"`
  Param6 string `json:"param6"`
  // ...
}

type SomeRequest struct {
	Query *SomeRequestQuery
	Path *SomeRequestPath
	Body *SomeRequestBody 
}

// ...
request := NewSomeRequest().WithBody(body).WithPath(path).WithQuery(query)

response, err := client.SomeApiRequest(request)
```

The response is a composite type

```
type SomeResponse struct {
  Response *http.Response
  Body []byte

  Payload *SomePayload // this is the default payload

  404Payload *404Payload
  400Payload *400Payload
  // ... whatever other payload is defined
}
```

That way we have access to the original response (think headers), to the response body and can also return different structs for specific status codes.  I'm not too sure currently on the different response types depending on the status code as it might add some complexity in deciding which payload is the default one and could potentially bloat the responses, depending on the swagger definition.

**Comment by arekkas** [comment]
By the way, I can definitely help with the Go code (review, help with templating), but I'm not experienced with Java developer. As a reference, I write a ton of Go and maintain popular libraries ([1](https://github.com/ory/hydra) [2](https://github.com/ory/ladon) [3](https://github.com/ory/dockertest]  [4](https://github.com/ory/fosite). Not intended as a brag but as confirmation that I am experienced with writing/maintaining Go libraries :)

**Comment by bvwells** [comment]
@arekkas many thanks for the comments on the proposed PR. I have to say I had some concerns about the  PR being a breaking change and also the body leak. Let's try to work out a sensible solution to the problem.

**Comment by antihax** [comment]
I like the idea of unmarshaling the error into a model and returning it as an abstract of the error type. We could take that further and return a []byte of the body as well as the model for cases such as if a CDN throws an identical status to what an operation would.

I will need to read up a bit more on how 3.0 would change parameters, it would be nice to keep required parameters in the functions parameter list and optionals as an additional struct handling `isSet` and `isNull`. 

IMO the library should handle if the parameter is body, query, or path if this is possible; not the developer. It would make life simpler than having to remember where each parameter goes.

**Comment by arekkas** [comment]
> IMO the library should handle if the parameter is body, query, or path if this is possible; not the developer. It would make life simpler than having to remember where each parameter goes.

I agree 100%. This can be solved with embedded structs:

```

type A struct {
  A string
  C string
}
type B struct {
  B string
  C string
}

type AB struct {
	A
	B
}

func main() {
	s := new(AB)
	s.A= "123"
	s.B = "123"

	// does not work: s.C = "123"
	s.A.C = "123"
	s.B.C = "123"
}
```

So here, the developer would only need to look this up if the names clash!

**Comment by antihax** [comment]
I would like to try to incorporate some of these changes to a library I am maintaining to see if it works but I have some work I need to complete first.

@arekkas @marians Does the following look ok for errors? https://play.golang.org/p/JC3LbPNkcaW
It would allow the error to be cast and expose both a []byte for the body and return the model, if one exists, for the error.

**Comment by arekkas** [comment]
Yes, that looks pretty good. I mean we still need to type assert it, but I think it's ok in this context and also somewhat standard (think stacktracer from [pkg/errors](https://github.com/pkg/errors/)).

The payload would be taken from the model for that response code, right?

ps: Which library?

**Comment by arekkas** [comment]
By the way, what do you estimate is the timeframe where this would get implemented? As stated before, I can definitely help with the Go code. We're also dependent on the 3.0 release of swagger-codegen (obviously). But to be blunt, the current approaches don't cut it because some cases just don't work - so I'd like to work towards a Go 3.0 release asap.

**Comment by bvwells** [comment]
Just wondering if anyones looked at the client code generated by go-swagger? Might be a source of inspiration. You can find an example generated client here:

https://github.com/go-swagger/go-swagger/tree/master/examples/task-tracker/client

**Comment by aeneasr** [comment]
What can I do to help bring this forward?

**Comment by webron** [comment]
@aeneasr the best way would be to submit a PR with the proposed changes, and hopefully others on this thread can review. We'll be happy to merge and pack it as part of the library.

**Comment by aeneasr** [comment]
I can help with API design and maybe some templating, but not with the java-portion. Any volunteers? :D

**Comment by Filippo125** [comment]
Hi, i tried to implement a better golang server generator with the parameter extraction programmatically and leave to developer only the controller function.
Please see my branch at [add_new_golang_server_generator](https://github.com/Filippo125/swagger-codegen-generators/tree/add_new_golang_server_generator). The generated code for pet store can be viewd at [GolangPetStore](https://github.com/Filippo125/GolangPetStore)

**Comment by giacomorock** [comment]
Hi! @Filippo125 seems a good point of start! Good Job!

**Comment by aeneasr** [comment]
Nice :) Just a side node, before touch client generation, take a look at this PR: https://github.com/OpenAPITools/openapi-generator/pull/8491
